module/os/linux/zfs/zpl_xattr.c

```diff
@@ -1329,6 +1330,660 @@ xattr_handler_t zpl_xattr_acl_default_handler =
 
 #endif /* CONFIG_FS_POSIX_ACL */
 
+#if defined(ZFS_NFS4_ACL)
+int
+zpl_permission(struct inode *ip, int mask)
+{
+	if (ITOZSB(ip)->z_acl_type == ZFS_ACLTYPE_NFS4ACL) {
+		/*
+		 * XXX - mask could also include
+		 * MAY_APPEND|MAY_ACCESS|MAY_OPEN|MAY_CHDIR, do we care?
```

Comment:
That's an excellent question which I hoped a reviewer with more detailed knowledge of the linux vfs implementation might have an opinion on :).

V_APPEND is a flag, which can be passed in the third argument to zfs_access. If this flag is set, and the initial permission checks in zfs_zaccess fail it calls zfs_zaccess_append. I think if the mask includes MAY_APPEND we could pass in the V_APPEND flag; but I'm not completely sure of the interaction. On the zfs side, I think it asks for write privs, which might fail, but then if you want write privs just for appending, it lets you through if the ACL allows ACE_APPEND_DATA. On the linux side, does it call with both MAY_WRITE and MAY_APPEND for an appended write or just MAY_APPEND?

Also, zfs_access in this use ends up calling zfs_zaccess_rwx which converts the mode bits mask to acl format bits and then calls zfs_zaccess. If it would work out better we could convert the modes bits and other flags into acl format bits directly in zpl_permission and bypass the conversion in zfs_zaccess_rwx by passing V_ACE_MASK in the flags for the call to zfs_access.

I just don't know enough about the linux vfs permission layer to say what's best at this point 8-/.